simulate-pcr: import from brewsci/science

suujia opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/brewsci/homebrew-bio/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/brewsci/homebrew-bio/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source FORMULA`, where `FORMULA` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict FORMULA` (after doing `brew install FORMULA`)?

-----

suujia commented:
mac OS  (but this link is fine):  
```
brewsci/bio/simulate-pcr:
  * Stable: The URL https://downloads.sourceforge.net/project/simulatepcr/simulate_PCR-v1.2.tar.gz is not reachable
```
both mac OS and linux: 
```
<255> expected but was <2>.
```

```
Exited with code 1
```

suujia commented:
Direct download link if necessary. (current link has a 5 second delay) 
``` 
https://downloads.sourceforge.net/project/simulatepcr/simulate_PCR-v1.2.tar.gz?r=https%3A%2F%2Fsourceforge.net%2Fprojects%2Fsimulatepcr%2Ffiles%2Fsimulate_PCR-v1.2.tar.gz%2Fdownload%3Fuse_mirror%3Dphoenixnap%26r%3Dhttps%253A%252F%252Fsourceforge.net%252Fprojects%252Fsimulatepcr%252Ffiles%252Fsimulate_PCR-v1.2.tar.gz%252Fdownload%253Fuse_mirror%253Dayera&ts=1525907781
```

tseemann commented:
I've had trouble with sourceforge https on macos too.
It eventually works.

tseemann commented:
Have a look `bioperl.rb` to see how you have to patch the Perl scripts to ensure Bioperl is found.

sjackman commented:
`simulate-pcr` depends on `LWP::Simple`. See for example `mlst` for how to add Perl dependencies to a formula.
https://github.com/brewsci/homebrew-bio/blob/f2cfb685f0be67b55601191b83fce12a71ed367a/Formula/mlst.rb#L26-L27

sjackman commented on Formula/simulate-pcr.rb:
```diff
@@ -7,8 +7,11 @@ class SimulatePcr < Formula
 
   def install
+    cmd = Language::Java.java_home_cmd("1.8")
+    ENV["JAVA_HOME"] = Utils.popen_read(cmd).chomp
```
I didn't think that `simulate-pcr` required Java.

sjackman commented on Formula/simulate-pcr.rb:
```diff
@@ -7,8 +7,11 @@ class SimulatePcr < Formula
 
   depends_on "bioperl"
   depends_on "blast"
+  depends_on "LWP::Simple" => :perl
```
This was the previous now defunct way of adding Perl dependencies. It sadly no longer works. See for example `mlst` for how to add Perl dependencies to a formula.
https://github.com/brewsci/homebrew-bio/blob/f2cfb685f0be67b55601191b83fce12a71ed367a/Formula/mlst.rb#L26-L27

suujia commented:
``` 
# Calling 'depends_on ... => :perl' is disabled!
``` 
change  `depends_on "LWP::Simple" => :perl` to `depends_on "perl" unless OS.mac?` 
(let me know if it's incorrect)

suujia commented:
``` 
Error: No such file or directory - simulate_PCR

```

suujia commented:
Simulate-pcr dependencies from their download website
```
LWP::Simple
Getopt::Long
Bio::SeqIO

makeblastdb, blastn, and blastdbcmd (http://blast.ncbi.nlm.nih.gov/) must be in your path. As of testing, BLAST should be version 2.2.27+ or higher, so that all options are available.
```

 after adding these dependencies i got error: 
``` 
! Installing the dependencies failed: Module 'Net::HTTP' is not installed
! Bailing out the installation for libwww-perl-6.34.
```

and after installing net::http as well: 
```
Building and testing Net-HTTP-6.18 ... ! Installing Net::HTTP failed. See /private/tmp/simulate-pcr-20180612-78875-4gdpzl/.brew_home/.cpanm/work/1528835434.78888/build.log for details. Retry with --force to force install it.
```
so then I installed net::http with "--force" (let me know if I shouldn't) - but I read that it just meant some of the tests failed, and this was a suggested solution [https://www.thegeekstuff.com/2013/06/cpan-force-install-perl-module/](url)

sjackman commented:
This PR is currently failing with this error:
```
Error: No such file or directory - simulate_PCR
```
https://travis-ci.org/brewsci/homebrew-bio/builds/380855431#L3336

Due to…
```ruby
    libexec.install Dir["*"]
    bin.install "simulate_PCR"
```
The first line installs (moves) all the files to `libexec`, and then the second line attempts to install `simulate_PCR`, which fails, because it's already been moved. You can delete the first `libexec.install` line.

suujia commented:
locally, without    ` libexec.install Dir["*"]`

i get the error message: 

```
Can't locate LWP/Simple.pm in @INC (you may need to install the LWP::Simple module) (@INC contains: /home/schen/.homebrew/Cellar/perl/5.26.2/lib/perl5/site_perl/5.26.2/darwin-thread-multi-2level /home/schen/.homebrew/Cellar/perl/5.26.2/lib/perl5/site_perl/5.26.2 /home/schen/.homebrew/Cellar/perl/5.26.2/lib/perl5/5.26.2/darwin-thread-multi-2level /home/schen/.homebrew/Cellar/perl/5.26.2/lib/perl5/5.26.2 /home/schen/.homebrew/lib/perl5/site_perl/5.26.2/darwin-thread-multi-2level /home/schen/.homebrew/lib/perl5/site_perl/5.26.2) at /home/schen/.homebrew/bin/simulate_PCR line 46.
BEGIN failed--compilation aborted at /home/schen/.homebrew/bin/simulate_PCR line 46.
```

sjackman commented:
I get this error locally on macOS.
```
Building and testing Net-HTTP-6.18 ... ! Installing Net::HTTP failed. See /private/tmp/simulate-pcr-20180615-2577-yam77w/.brew_home/.cpanm/work/1529087023.2586/build.log for details. Retry with --force to force install it.
FAIL
```
In `build.log`
```
t/00-report-prereqs.t .. ok
t/http-nb.t ............ ok
t/http.t ............... ok
Net::HTTPS: SSL connect attempt failed with unknown error error:1407742E:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert protocol version at t/live-https.t line 34.
# Looks like your test exited with 255 before it could output anything.
t/live-https.t ......... 
Dubious, test returned 255 (wstat 65280, 0xff00)
Failed 6/6 subtests 
t/live.t ............... ok
```
https://gist.github.com/sjackman/775b70b79cae458f08373fdbf3fef55b#file-build-log-L1457

Googling that error message gives these two hits:
https://rt.cpan.org/Public/Bug/Display.html?id=123583
https://github.com/libwww-perl/Net-HTTP/issues/47
In the latter issue they say that upgrading IO::Socket::SSL helps.

The following patch fixes the issue locally for me:
```diff
--- a/Formula/simulate-pcr.rb
+++ b/Formula/simulate-pcr.rb
@@ -15,7 +15,7 @@ class SimulatePcr < Formula
     bin.install "simulate_PCR"
     inreplace bin/"simulate_PCR", "#!/usr/bin/perl", "#!/usr/bin/env perl"
     ENV.prepend_create_path "PERL5LIB", libexec/"lib/perl5"
-    system "cpanm", "--self-contained", "-l", libexec, "LWP::Simple"
+    system "cpanm", "--self-contained", "-l", libexec, "IO::Socket::SSL", "LWP:
     (bin/"simulate-pcr").write_env_script("#{libexec}/bin/simulate-pcr", :PERL5
   end
```

sjackman commented:
To fix this error:
```
Can't locate LWP/Simple.pm in @INC (you may need to install the LWP::Simple module) (@INC contains: /home/schen/.homebrew/Cellar/perl/5.26.2/lib/perl5/site_perl/5.26.2/darwin-thread-multi-2level /home/schen/.homebrew/Cellar/perl/5.26.2/lib/perl5/site_perl/5.26.2 /home/schen/.homebrew/Cellar/perl/5.26.2/lib/perl5/5.26.2/darwin-thread-multi-2level /home/schen/.homebrew/Cellar/perl/5.26.2/lib/perl5/5.26.2 /home/schen/.homebrew/lib/perl5/site_perl/5.26.2/darwin-thread-multi-2level /home/schen/.homebrew/lib/perl5/site_perl/5.26.2) at /home/schen/.homebrew/bin/simulate_PCR line 46.
BEGIN failed--compilation aborted at /home/schen/.homebrew/bin/simulate_PCR line 46.
```
try this change
```diff
diff --git a/Formula/simulate-pcr.rb b/Formula/simulate-pcr.rb
index 448fd7b..5c63631 100644
--- a/Formula/simulate-pcr.rb
+++ b/Formula/simulate-pcr.rb
@@ -16,7 +16,7 @@ class SimulatePcr < Formula
     inreplace bin/"simulate_PCR", "#!/usr/bin/perl", "#!/usr/bin/env perl"
     ENV.prepend_create_path "PERL5LIB", libexec/"lib/perl5"
     system "cpanm", "--self-contained", "-l", libexec, "LWP::Simple"
-    (bin/"simulate-pcr").write_env_script("#{libexec}/bin/simulate-pcr", :PERL5LIB => ENV["PERL5LIB"])
+    bin.env_script_all_files libexec, "PERL5LIB" => ENV["PERL5LIB"]
   end
 
   test do
```

suujia commented:
produced locally: 
``` 
Can't locate Bio/SeqIO.pm in @INC (you may need to install the Bio::SeqIO module) (@INC contains: /home/linuxbrew/.linuxbrew/Cellar/perl/5.26.2/lib/perl5/site_perl/5.26.2/x86_64-linux-thread-multi /home/linuxbrew/.linuxbrew/Cellar/perl/5.26.2/lib/perl5/site_perl/5.26.2 /home/linuxbrew/.linuxbrew/Cellar/perl/5.26.2/lib/perl5/5.26.2/x86_64-linux-thread-multi /home/linuxbrew/.linuxbrew/Cellar/perl/5.26.2/lib/perl5/5.26.2 /home/linuxbrew/.linuxbrew/lib/perl5/site_perl/5.26.2/x86_64-linux-thread-multi /home/linuxbrew/.linuxbrew/lib/perl5/site_perl/5.26.2) at /home/linuxbrew/.linuxbrew/bin/simulate_PCR line 48.
```

sjackman commented:
> produced locally:
`Can't locate Bio/SeqIO.pm in @INC (you may need to install the Bio::SeqIO module)`
added:
`system "cpanm", "--self-contained", "-l", libexec, "Bio::SeqIO"`

Bio::SeqIO is provided by `bioperl`. You can fix this error with…

```diff
+    ENV.prepend "PERL5LIB", Formula["bioperl"].libexec/"lib/perl5"
     ENV.prepend_create_path "PERL5LIB", libexec/"lib/perl5"
```

It should work after that!

sjackman commented:
💚 Woohoo!

sjackman commented on Formula/simulate-pcr.rb:
```diff
@@ -0,0 +1,26 @@
+  # cite Gardner_2014: "https://doi.org/10.1186/1471-2105-15-237"
+  desc "Predicts amplicon products from single or multiplex primers"
+  homepage "https://sourceforge.net/projects/simulatepcr/"
+
```
Please delete this blank line.

sjackman commented:
Excellent! Please squash the 12 commits down to a  single commit. I use `git rebase -i origin/master` for that purpose. Let me know if you need a hand with that.

suujia commented:
My bad, I was editing everything on web interface and couldn't find it locally. Noop popped up trying to rebase in my remote and origin. So I moved the head to my latest commit and squashed from there. Then by force pushing to override this PR, it closed it due to the current head being not a descendant of the stored head sha. continuation: 
https://github.com/brewsci/homebrew-bio/pull/320

tseemann commented:
@suujia does the tool `isPcr` do what you need? It's in brew already.

sjackman commented:
Hi, Torsten. We're migrating `simulate-pcr` because it's in BCGSC's ORCA docker image, and `brew test simulate-pcr` was failing. I see that you originally added `simulate-pcr` over at PR https://github.com/homebrew/homebrew-science/pull/2334. Which tool do you tend to use, `simulate-pcr` or `ispcr`?
